### DISCUSSION TITLE
mesa-frames: faster mesa with data-frame based operations

### DISCUSSION THREAD

**Discussion by adamamer20** [mesa/mesa#1939] [Show and tell]
Hello,
I am excited to introduce a new extension for the Mesa simulation library, named [mesa-frames](https://github.com/adamamer20/mesa-frames). This extension is designed to enhance the performance of models with thousands of agents by leveraging pandas dataframes for storing agent attributes and utilizing vectorized operations. In cases where vectorized operations are not feasible, operations can be performed by row by using .apply. This approach has demonstrated greater efficiency than parallel processing the steps of the agents in my [project](https://github.com/adamamer20/streetcrime).
The goal of mesa-frames is to facilitate an easy transition from the base Mesa library, ensuring compatibility with existing Mesa operations while introducing the capability to 'vectorize' a model using a decorator. Currently, the extension is in its initial proof-of-concept stage.
I warmly welcome feedback, suggestions, and contributions to this project. 
Thank you.

**Comment by Corvince** [comment]
Wow that sounds super interesting and the performance chart looks very promising. 

I am curious what kind of problems can be modeled this way and how we can facilitate the interoperability with mesa. 

Congratulations on the launch

**Comment by EwoutH** [comment]
Looks awesome, really excited to see where it’s going!

If you need any (specific) input, feel free to tag me.

**Comment by rht** [comment]
Ran the benchmark for small number of agents.
![small_number_agents](https://github.com/projectmesa/mesa/assets/395821/40ed31e4-6e30-45c1-9c4f-017c00d3c906)
The upfront baseline runtime of 0.25 s is worth the tradeoff of being slower than the current `AgentSet` implementation. I think we can mold the API of mesa-frames to have the methods `AgentSet` (`select`, `shuffle`, `sort`, `do`). Or alternatively, replace the data structure backend of `AgentSet` from `weakref.WeakKeyDictionary()` to a pandas DF.

Edit: clarification

**Reply by rht** [reply]
@EwoutH this seems to be something you can experiment with quickly the efficacy of.

**Reply by adamamer20** [reply]
This seems like a good idea!
However the `do` method might be problematic (and this is also related to "vectorizing" an existing model).
The use case I envisioned with mesa-frames was to run a model with `SimultaneousActivation` (or a method where the agents states were independent) faster.
However the only way I was able to achieve those speeds is by using built-in pandas vectorized operations. If i were to use a custom method I could only use `pd.apply` row-wise, it would probably be slower than iterating through a list.
Thus for people unfamiliar with pandas API or those "vectorizing" an existing model, the `do` method would be slower.
There are some solutions I was considering: The first was switching to polars as the `map_rows` implementation probably acts in parallel with respect to pandas. The other was using `swifter` which attempts to vectorize a method and fall back to parralelizing with `dusk` when this isn't possible. I will try to run some tests soon.

**Reply by EwoutH** [reply]
What happens in a model that has (large amount) of creating and removal of agents during runtime? Is it compatible? Is it fast?

**Reply by rht** [reply]
I personally think that proper models need to strive to always have `SimultaneousActivation`/`StagedActivation` in mind. Because a random activation would imply first mover advantage in the agents that `step()` first. So I still think you can do general modeling with mesa-frames, with a vectorized `do`. An example of this would be a parallelized implementation of the classic Sugarscape model, that allows it to run on GPU. Relevant paper: [Who Goes First?
An Examination of the Impact of Activation on Outcome Behavior in Agent-based Models](mars.gmu.edu/bitstream/handle/1920/9070/Comer_gmu_0883E_10539.pdf).

**Reply by rht** [reply]
> What happens in a model that has (large amount) of creating and removal of agents during runtime? Is it compatible? Is it fast?

This has a builtin operation in pandas, `df.drop(df[df.A > 3].index)`, and could be translated to the ABM context.

**Reply by rht** [reply]
@adamamer20 you can check out https://github.com/projectmesa/mesa-examples/pull/79 for a showcase of the new AgentSet API.

**Reply by adamamer20** [reply]
Thank you! It definitely seems implementable using pandas without any issues and seems clearer than using the scheduler. Maybe on the documentation it can be added that one should use builtin functions and only in the case where it is not possible using a custom function.
However I still think some models might require sequential activations like `RandomActivation` (e.g. a model of a stock market where price changes after the trading decision of an agent or strategic interactions in evolutionary games). Maybe adding an option to `do` which specify sequentiality or simultaneity?

**Reply by rht** [reply]
> However I still think some models might require sequential activations like RandomActivation (e.g. a model of a stock market where price changes after the trading decision of an agent or strategic interactions in evolutionary games). Maybe adding an option to do which specify sequentiality or simultaneity?

That's true, in the case when it is unavoidable to do `RandomActivation`. I suppose having a `do` method that applies on a shuffled rows sequentially, would be convenient, for feature completeness sake, albeit slow. I think you could try PyArrow as the [backend](https://pandas.pydata.org/docs/user_guide/pyarrow.html) to pandas instead of the NumPy backend, which may be even faster.

**Reply by EwoutH** [reply]
Wondering if any kind of queuing or pipelining can help speed this up. You can shuffle once, and then the order and the operations are known, only the exact input data isn’t (since previous agent actions may have changed states).

**Edit:** with the exception of agents being removed by other agents.

**Reply by rht** [reply]
Using pola.rs
![polars](https://github.com/projectmesa/mesa/assets/395821/b7aa3729-e11c-4a6d-8309-4173e88efedd)
with the following DF code
```python
class MoneyModelDF(ModelDF):
    def __init__(self, N):
        super().__init__()
        self.num_agents = N
        self.agents = pl.DataFrame({"wealth": [1 for i in range(N)]})

    def step(self, merged_mro=True):
        wealthy_agents = self.agents.filter(pl.col("wealth") > 0)
        n_wealthy = len(wealthy_agents)
        if n_wealthy > 0:
            self.agents = self.agents.with_columns(wealth=pl.when(pl.col("wealth") > 0).then(pl.col("wealth") - 1).otherwise("wealth"))

            rows_to_update = np.random.choice(self.agents.height, size=n_wealthy, replace=False)
            # Could be wrong, but this is not going to be much slower than the previous with_columns operation anyway
            self.agents = self.agents.with_columns(wealth=pl.when(pl.col("wealth").is_in(rows_to_update)).then(pl.col("wealth") + 1).otherwise("wealth"))
```

**Reply by rht** [reply]
With Polars, we can get a huge speedup (~10x over pandas) not just for the AgentSet, but also on agent `step()`. However, Polars native expression API is hard to translate to how users would usually define their agent `step()`. There are definitely lots of use case of ABM that is not tied to geopandas (e.g. economic models). As such, I think it is worth implementing a variant of mesa-frames using Polars.

**Reply by adamamer20** [reply]
The speedup is amazing! I initially started with subclassing Dataframes but now i have switched to using a custom accessor to avoid the complexity of constructors. Since also polars has a similar api for extensions, i think it wouldn't be much more time consuming doing both and it is definetely worth it.

**Reply by rht** [reply]
I didn't believe it at first. Too good to be true. But the earlier versions of my code resembled idiomatic pandas, and was slightly slower than the pandas version. There were warnings saying that it wasn't using Polars native expression API, and I fixed them one by one until I got to the version I posted above.

**Reply by rht** [reply]
Context: https://github.com/projectmesa/mesa/discussions/1561, cc: @mrgoodfish check out the Polars vs pandas benchmark result above.

**Comment by adamamer20** [comment]
> That's true, in the case when it is unavoidable to do `RandomActivation`. I suppose having a `do` method that applies on a shuffled rows sequentially, would be convenient, for feature completeness sake, albeit slow. I think you could try PyArrow as the [backend](https://pandas.pydata.org/docs/user_guide/pyarrow.html) to pandas instead of the NumPy backend, which may be even faster.

Sure, I will try to test both.

> Wondering if any kind of queuing or pipelining can help speed this up. You can shuffle once, and then the order and the operations are known, only the exact input data isn’t (since previous agent actions may have changed states).
> 
> **Edit:** with the exception of agents being removed by other agents.

if the operation relies on the states of x previous agents one could use pandas [rolling](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html) method. In this case, the implementation should still be left to the user. Otherwise, if the operation relies on external states or, as you mentioned, agents are removed, I do not see any other way of doing this then iterating sequentially through the rows.

I will try to implement AgentSet using dataframes soon and let you know.

**Reply by rht** [reply]
See also https://pola.rs/ for performance.

**Reply by rht** [reply]
@adamamer20 is it correct that in order to test for the PyArrow backend, in https://github.com/adamamer20/mesa-frames/blob/9bee978dc29a9751d9c7fcc3b6c2ca614a4b2da6/mesa_frames/model.py#L262, I added `"[pyarrow]"`?

**Reply by adamamer20** [reply]
> @adamamer20 is it correct that in order to test for the PyArrow backend, in https://github.com/adamamer20/mesa-frames/blob/9bee978dc29a9751d9c7fcc3b6c2ca614a4b2da6/mesa_frames/model.py#L262, I added `"[pyarrow]"`?

yes, that's correct. i believe that in that case there's no need to convert dtype=int to dtype=float because I remember reading somewhere that PyArrow supports missing values with int.

As for polars, it is definitely worth to try because it allows also to use  a dataset larger than the RAM (useful if you have millions of agents). The only two potential issues was a slightly different interface than pandas and the lack of pandas extensions (for example, it's missing  the equivalent of geopandas which is necessary for `mesa-geo`).

**Reply by rht** [reply]
@adamamer20 have you seen https://github.com/geopolars/geopolars ? I found it via https://github.com/cnpryer/awesome-polars.

**Reply by adamamer20** [reply]
Nice! Seems very promising. Towards the end of GSOC, we can try building GeoAgentSets

**Comment by jackiekazil** [comment]
This is exciting stuff!

**Comment by adamamer20** [comment]
@rht @Corvince @EwoutH @jackiekazil let me know what you think of the implementation so far in [adamamer20/mesa_frames#8](https://github.com/adamamer20/mesa-frames/pull/8). If you like the API, I will implement the backend with polars.

**Reply by rht** [reply]
The API LGTM. Do you want to do this as a Google Summer of Code project? A medium project length could be 175 hours in total. A hard project length could be 350 hours.

The deadline, since Mesa participates in GSoC via NumFOCUS (see https://github.com/numfocus/gsoc/blob/master/2024/ideas-list.md):
> NumFOCUS Deadline: February 5 at 1800 UTC (24 hours before the Google GSOC deadline on February 6 at 1800 UTC)

@tpike3 can help in writing the proposal.

**Reply by adamamer20** [reply]
Wow that sounds amazing! I think it would be a good medium project. How do I proceed?

**Reply by tpike3** [reply]
@adamamer20 We put in for the GSOC 2024 if we get accepted then you apply to be a contributor to it. https://summerofcode.withgoogle.com/ If we dont get accepted, then we can put in for a Small Development grant through Numfocus.

**Reply by rht** [reply]
@adamamer20  check out https://github.com/projectmesa/mesa/discussions/1997 and https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2024.

**Reply by adamamer20** [reply]
@tpike3 @rht thank you very much for the project proposal!

**Reply by rht** [reply]
> We put in for the GSOC 2024 if we get accepted then you apply to be a contributor to it.

@adamamer20 Mesa has been accepted: https://summerofcode.withgoogle.com/programs/2024/organizations/project-mesa. The application opens on March 18.

**Reply by adamamer20** [reply]
@rht wonderful news, thank you. I will apply as soon as it opens.

**Reply by rht** [reply]
@adamamer20 we haven't heard back from you. Let's continue the discussion on https://matrix.to/#/#project-mesa:matrix.org.

**Comment by EwoutH** [comment]
@adamamer20 Are you still working on this? I might want to run some simulations with around ~100.000 agents so I might become very interested - and invested - in this.

**Reply by adamamer20** [reply]
@EwoutH Sure! I was waiting on the results and feedback from the GSoC proposal before going back into development. If you want to provide some guidance or contribute, it would be more than welcome. What are your priorities?

**Reply by EwoutH** [reply]
Great to hear! I'm going to dive deeper in it in the coming week. I will be simulating a city of a hundred thousand to a million inhabitants, but with limited interactivity and relative simple behaviour.

I would like mesa-frames to be a clean subset of mesa. Maybe we can integrate it in the mesa organization, like we have done with Mesa-geo. I will take this up with the maintainers.

Could you check out https://github.com/projectmesa/mesa/pull/2066? One of the things I would like to be able to do, is efficiently simulate a day, in which agents do only one or a few times something, while still having a relatively high temporal resolution (minutes). Discrete event scheduling would be perfect for that. But that might be relatively difficult with frames (or not, it's a gut feeling).

As of GSoC, we would have to look unbiased of course, but we certainly consider it. In any case I hope we can keep developing mesa-frames.

**Reply by adamamer20** [reply]
That's the direction I intended for mesa-frames too!
I think supporting event scheduling shouldn't be too difficult. I have to dive deeper into the code of #2066 but adding a column for events and masking active_agents on that column should work. I will try to add something like it in the coming weeks.
Regardless of the GSoC outcome, I'm commited to mesa-frames but the majority of the development will still take place during summer. With courses and exams, the time I can dedicate to it is limited.

**Comment by adamamer20** [comment]
@jackiekazil @tpike3 @rht @wang-boyu @quaquel @Corvince @EwoutH  
Thank you all for accepting my proposal!
In the next weeks I'll develop a more concrete plan but I plan on closing adamamer20/mesa-frames#8 soon and work on adamamer20/mesa-frames#9. I'll also familiarize myself with the latest 3.0.0 developments. I'll use this as the main discussion thread.

**Reply by rht** [reply]
Congratulations, @adamamer20. We will also have to think of a stable API so that maintaining mesa-frames is more manageable. And visualization integration.

**Reply by adamamer20** [reply]
For the API, I was experimenting with dunder methods to align to pandas/polars syntax and make it more user-friendly. I think they could also be added to mesa.AgentSet.

**Comment by rht** [comment]
Regarding with the naming, I'm wondering if it is easier to remember to name it mesa-frame instead of mesa-frames? I don't see a rationale for making it plural.

**Reply by adamamer20** [reply]
I'm not opposed to that. I was also dubious about the hyphen because python beginners may find it confusing (you need to install mesa-frame but import mesa_frame)

**Reply by jackiekazil** [reply]
I thought it was frames, because it involved dataframes (plural). I don't care enough about "s" or no "s". So I defer to you all. 

RE: hyphen versus underscore - this is long been a thing in python of inconsistency. You can see this happening in our community as well -- names with hyphens (mesa-geo), names which are two words together (cropyield).  

Some notes on this --  
* package name uses hyphen if needed, because SEO, I believe - https://www.stanventures.com/blog/underscore-in-url/#:~:text=Google%20recommends%20using%20hyphens%20rather,find%20in%20Google's%20official%20documentation. 
* underscores, while possible are discouraged according to pep8 - https://peps.python.org/pep-0008/#package-and-module-names

These are the things that end up being something that we (engineers) can never agree on, but spend a lot of time talking about. 
Now that I have gotten to the end of my comment, I am not sure any of it was really helpful.
